src/v/archival/utils.h

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/vectorizedio/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+
+#include "seastarx.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/gate.hh>
+#include <seastar/core/semaphore.hh>
+
+#include <memory>
+
+namespace archival {
+
+/// \brief RAII wrapper that holds the door^W gate open
+///        until the end of scope
+///
+/// The guard is moveable. The lifetime will be transferred
+/// to the new gate_guard instance.
+///
+/// \example ss::gate gate;
+///          {
+///             gate_guard g{gate};
+///             assert(gate.get_count() == 1);
+///          }
+///          assert(gate.get_count() == 0);
+struct gate_guard {
```

Comment:
i think this can be a separate commit and go into `src/utils` as a general utility?